drivers/adc/adc_stm32f4.c

```diff
@@ -0,0 +1,501 @@
+	GPIOport =  NULL;
+
+	switch (c) {
+		case adcChannel_0:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_0;
+			cfg.Channel = ADC_CHANNEL_0;
+			break;
+		case adcChannel_1:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_1;
+			cfg.Channel = ADC_CHANNEL_1;
+			break;
+		case adcChannel_2:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_2;
+			cfg.Channel = ADC_CHANNEL_2;
+			break;
+		case adcChannel_3:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_3;
+			cfg.Channel = ADC_CHANNEL_3;
+			break;
+		case adcChannel_4:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_4;
+			cfg.Channel = ADC_CHANNEL_4;
+			break;
+		case adcChannel_5:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_5;
+			cfg.Channel = ADC_CHANNEL_5;
+			break;
+		case adcChannel_6:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_6;
+			cfg.Channel = ADC_CHANNEL_6;
+			break;
+		case adcChannel_7:
+			GPIOport = GPIOA;
+			gpioCfg.Pin = GPIO_PIN_7;
+			cfg.Channel = ADC_CHANNEL_7;
+			break;
+		case adcChannel_8:
+			GPIOport = GPIOB;
+			gpioCfg.Pin = GPIO_PIN_0;
+			cfg.Channel = ADC_CHANNEL_8;
+			break;
+		case adcChannel_9:
+			GPIOport = GPIOB;
+			gpioCfg.Pin = GPIO_PIN_1;
+			cfg.Channel = ADC_CHANNEL_9;
+			break;
+		case adcChannel_10:
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_0;
+			cfg.Channel = ADC_CHANNEL_10;
+			break;
+		case adcChannel_11:
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_1;
+			cfg.Channel = ADC_CHANNEL_11;
+			break;
+		case adcChannel_12:
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_2;
+			cfg.Channel = ADC_CHANNEL_12;
+			break;
+		case adcChannel_13:
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_3;
+			cfg.Channel = ADC_CHANNEL_13;
+			break;
+		case adcChannel_14:
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_4;
+			cfg.Channel = ADC_CHANNEL_14;
+			break;
+		case adcChannel_15:
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_5;
+			cfg.Channel = ADC_CHANNEL_15;
+			break;
+
+		case adcChannel_vref:
+			cfg.Channel = ADC_CHANNEL_VREFINT;
+			ADC->CCR |= ADC_CCR_TSVREFE;
+			break;
+
+		case adcChannel_temp:
+			cfg.Channel = ADC_CHANNEL_TEMPSENSOR;
+			break;
+
+		case adcChannel_vbat:
+			cfg.Channel = ADC_CHANNEL_VBAT;
+			ADC->CCR |= ADC_CCR_VBATE;
+			break;
+
+		default:
+			// none
+			break;
+	}
+	if(GPIOport) HAL_GPIO_Init(GPIOport, &gpioCfg);
+
+	if (HAL_ADC_ConfigChannel(phadc, &cfg) != HAL_OK) {
+		SYS_LOG_ERR("config adc channel failed");
+		return stm32adcError_adcHALconfigChannel;
+	}
+	SYS_LOG_DBG("internal adc channel %u has been configured",cfg.Channel);
+
+	return stm32adcError_None;
+}
+
+int adc_stm32f4_init(struct device *dev)
+{
+	const struct adc_config *config = dev->config->config_info;
+	struct adc_drvData *drvData = dev->driver_data;
+	adc_channel_index_t i;
+	uint32_t r;
+	uint32_t activeChannels;
+
+	SYS_LOG_INF("init adc%u", (unsigned int)config->adcDevNum);
+
+	switch(config->adcDevNum) {
+		#ifdef CONFIG_ADC_0
+		case 0:
+			__HAL_RCC_ADC1_CLK_ENABLE();
```

Comment:
The best think would be to use a "clock_control.h" etc. instead of the bare HAL call. But I do not know how to connect to other drivers and what are the interfaces Zephyr is providing for that.

The ADC clock can be generated in two ways. I only implemented one as a straight forward solution